### PR TITLE
[LibOS] Make async worker persistent

### PR DESCRIPTION
**Pull request by mwkmwkmwk**
## Description of the changes 

The async worker thread is currently spawned on-demand and terminated after an idle timeout. This is entirely unnecessary — the cost of having a persistent async thread is near-0, and it simplifies the logic.

Fixes gramineproject/graphene#440.

## How to test this PR? 

Already covered by CI.

- - -
This change is [Reviewable](https://reviewable.io/reviews/gramineproject/gramine/1992)

### CODE REVIEW

**Review by mkow — CHANGES_REQUESTED** [review]
> +28 −98

nice!

> Reviewed 3 of 3 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/1992#-:-O5uerFK7gnoleNV_ehf:bzkrpl)* status: all files reviewed, 1 unresolved discussion, not enough approvals from maintainers (3 more required), not enough approvals from different teams (2 more required, approved so far: ) (waiting on @mwkmwkmwk)

___
*[`-- commits` line 8 at r1](https://reviewable.io/reviews/gramineproject/gramine/1992#-O5ubnbdB5sp4xRASf9C):*
Please briefly describe what was that bug here (just a single sentence should be enough), we want to keep the git repo standalone and self-sufficient, in case we ever move to some other repo hosting.
And if you want to include a link to the GitHub issue, please use a full HTTPS URL.

**Review by mwkmwkmwk — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/1992)* status: all files reviewed, 1 unresolved discussion, not enough approvals from maintainers (3 more required), not enough approvals from different teams (2 more required, approved so far: ) (waiting on @mkow)

___
*[`-- commits` line 8 at r1](https://reviewable.io/reviews/gramineproject/gramine/1992#-O5ubnbdB5sp4xRASf9C):*
<details><summary><i>Previously, mkow (Michał Kowalczyk) wrote…</i></summary><blockquote>

Please briefly describe what was that bug here (just a single sentence should be enough), we want to keep the git repo standalone and self-sufficient, in case we ever move to some other repo hosting.
And if you want to include a link to the GitHub issue, please use a full HTTPS URL.
</blockquote></details>

Done.

**Review by dimakuv — CHANGES_REQUESTED** [review]
> Reviewed 3 of 3 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/1992)* status: all files reviewed, 6 unresolved discussions, not enough approvals from maintainers (3 more required), not enough approvals from different teams (2 more required, approved so far: ) (waiting on @mkow and @mwkmwkmwk)

*[a discussion](https://reviewable.io/reviews/gramineproject/gramine/1992#-O5vrq8R5S0qQkPG029v:-O5vrq8R5S0qQkPG029w:b-9m3zp1) (no related file):*
FYI: I think this code was left from the times when helper threads (including this Async helper thread) had a spinning implementation of waiting on events. Of course now with proper `PalStreamsWaitEvents()` that puts the thread to sleep on the host, this "terminate the thread when no events" logic is not needed.

Fun times of barely working code from 2010s...

___
___
*[`-- commits` line 4 at r2](https://reviewable.io/reviews/gramineproject/gramine/1992#-O5vogjT1YwoTwZlGMp1):*
Typically, we write the current status in past tense, like `Previously, the async worker thread spawned ...`. Should we continue with this style? Currently it sounds a bit weird -- the commit message is supposed to describe the new status (imho), but now it describes the old status.
___
*[`libos/src/libos_async.c` line 29 at r2](https://reviewable.io/reviews/gramineproject/gramine/1992#-O5vqZKL-_u8rtsZauX-:-O5vqZKL-_u8rtsZauX0:b-lpbtgv) ([raw file](https://github.com/gramineproject/gramine/blob/180dbc57ccabb2c18afd83925df9da45776bc8c2/libos/src/libos_async.c#L29)):*
> ```C
> static LISTP_TYPE(async_event) async_list;
> 
> /* Should be accessed with async_worker_lock held. */
> ```

1. This comment is placed in a way that makes you think that `async_worker_running` must also be accessed with the lock held -- this is not true.
2. Also, do we even need the lock? Why not just use atomics for `async_worker_shutdown`? We'll save a couple more lines of code then.
___
*[`libos/src/libos_async.c` line 37 at r2](https://reviewable.io/reviews/gramineproject/gramine/1992#-O5vrV9rA4e_fojTHAeC:-O5vrV9rA4e_fojTHAeD:b-xlziev) ([raw file](https://github.com/gramineproject/gramine/blob/180dbc57ccabb2c18afd83925df9da45776bc8c2/libos/src/libos_async.c#L37)):*
> ```C
> 
> /* TODO: use async_worker_thread->pollable_event instead */
> static struct libos_pollable_event install_new_event;
> ```

I would also prefer to add the prefix `g_` to all these global variables (in a separate PR), but I won't block on this.
___
*[`libos/src/libos_async.c` line 327 at r2](https://reviewable.io/reviews/gramineproject/gramine/1992#-O5vqDEdAhj_Kf42qwj5:-O5vqDEdAhj_Kf42qwj6:b2sutwz) ([raw file](https://github.com/gramineproject/gramine/blob/180dbc57ccabb2c18afd83925df9da45776bc8c2/libos/src/libos_async.c#L327)):*
> ```C
>         return -ENOMEM;
> 
>     async_worker_thread = new;
> ```

Can we move this global assignment to the very end of the func? This will be more in line with our conventions.
___
*[`libos/src/libos_async.c` line 356 at r2](https://reviewable.io/reviews/gramineproject/gramine/1992#-O5vrBYAEukl6XrrHKn9:-O5vrBYB1Rk_MOYGsfvN:b-k6xu5l) ([raw file](https://github.com/gramineproject/gramine/blob/180dbc57ccabb2c18afd83925df9da45776bc8c2/libos/src/libos_async.c#L356)):*
> ```C
>     put_thread(async_worker_thread);
>     async_worker_thread = NULL;
>     PalObjectDestroy(handle);
> ```

I think we don't need these four lines of clean-up. This happens on process termination, so all resources will be cleaned up by the host kernel anyway.

Also, if you want to do clean-up, then you should also destroy the lock `async_worker_lock` and destroy `install_new_event`. Right now you have something in-between...

**Review by mkow — CHANGES_REQUESTED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/1992)* status: all files reviewed, 5 unresolved discussions, not enough approvals from maintainers (3 more required), not enough approvals from different teams (2 more required, approved so far: ) (waiting on @dimakuv and @mwkmwkmwk)

___
*[`-- commits` line 4 at r2](https://reviewable.io/reviews/gramineproject/gramine/1992#-O5vogjT1YwoTwZlGMp1):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

Typically, we write the current status in past tense, like `Previously, the async worker thread spawned ...`. Should we continue with this style? Currently it sounds a bit weird -- the commit message is supposed to describe the new status (imho), but now it describes the old status.
</blockquote></details>

Yup, +1, I'd say "previously", otherwise it's not obvious whether it describes the state before or after this commit.
___
*[`libos/src/libos_async.c` line 29 at r2](https://reviewable.io/reviews/gramineproject/gramine/1992#-O5vqZKL-_u8rtsZauX-:-O5wkQx61GXXRcoPJrHd:b-p4oilf) ([raw file](https://github.com/gramineproject/gramine/blob/180dbc57ccabb2c18afd83925df9da45776bc8c2/libos/src/libos_async.c#L29)):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

1. This comment is placed in a way that makes you think that `async_worker_running` must also be accessed with the lock held -- this is not true.
2. Also, do we even need the lock? Why not just use atomics for `async_worker_shutdown`? We'll save a couple more lines of code then.
</blockquote></details>

1. Let's move it to a trailing comment at the end of the line then?
2. Yeah, I guess we could. Whatever from my side, will leave the decision to you :)
___
*[`libos/src/libos_async.c` line 37 at r2](https://reviewable.io/reviews/gramineproject/gramine/1992#-O5vrV9rA4e_fojTHAeC:-O5wkiHR2uMGeG0CpNrc:bstvnax) ([raw file](https://github.com/gramineproject/gramine/blob/180dbc57ccabb2c18afd83925df9da45776bc8c2/libos/src/libos_async.c#L37)):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

I would also prefer to add the prefix `g_` to all these global variables (in a separate PR), but I won't block on this.
</blockquote></details>

Yeah, let's fix it separately.
___
*[`libos/src/libos_async.c` line 356 at r2](https://reviewable.io/reviews/gramineproject/gramine/1992#-O5vrBYAEukl6XrrHKn9:-O5wpuly56GmAh4MRoNI:baz1i12) ([raw file](https://github.com/gramineproject/gramine/blob/180dbc57ccabb2c18afd83925df9da45776bc8c2/libos/src/libos_async.c#L356)):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

I think we don't need these four lines of clean-up. This happens on process termination, so all resources will be cleaned up by the host kernel anyway.

Also, if you want to do clean-up, then you should also destroy the lock `async_worker_lock` and destroy `install_new_event`. Right now you have something in-between...
</blockquote></details>

Whatever @mwkmwkmwk decides, if choosing the first option, please add a comment explaining that this function is called when terminating the whole process so it doesn't need to clean up everything.

**Review by mwkmwkmwk — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/1992)* status: all files reviewed, 4 unresolved discussions, not enough approvals from maintainers (3 more required), not enough approvals from different teams (2 more required, approved so far: ) (waiting on @dimakuv and @mkow)

___
*[`-- commits` line 4 at r2](https://reviewable.io/reviews/gramineproject/gramine/1992#-O5vogjT1YwoTwZlGMp1):*
<details><summary><i>Previously, mkow (Michał Kowalczyk) wrote…</i></summary><blockquote>

Yup, +1, I'd say "previously", otherwise it's not obvious whether it describes the state before or after this commit.
</blockquote></details>

Done.

**Review by mkow — COMMENTED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/1992)* status: all files reviewed, 4 unresolved discussions, not enough approvals from maintainers (3 more required), not enough approvals from different teams (2 more required, approved so far: ) (waiting on @dimakuv and @mwkmwkmwk)

**Review by mwkmwkmwk — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/1992)* status: all files reviewed, 4 unresolved discussions, not enough approvals from maintainers (3 more required), not enough approvals from different teams (2 more required, approved so far: ) (waiting on @dimakuv and @mkow)

___
*[`libos/src/libos_async.c` line 29 at r2](https://reviewable.io/reviews/gramineproject/gramine/1992#-O5vqZKL-_u8rtsZauX-:-O61X6Qq7CqsvO20Zt1i:br0imz5) ([raw file](https://github.com/gramineproject/gramine/blob/180dbc57ccabb2c18afd83925df9da45776bc8c2/libos/src/libos_async.c#L29)):*
<details><summary><i>Previously, mkow (Michał Kowalczyk) wrote…</i></summary><blockquote>

1. Let's move it to a trailing comment at the end of the line then?
2. Yeah, I guess we could. Whatever from my side, will leave the decision to you :)
</blockquote></details>

removed the lock in favor of an atomic
___
*[`libos/src/libos_async.c` line 327 at r2](https://reviewable.io/reviews/gramineproject/gramine/1992#-O5vqDEdAhj_Kf42qwj5:-O61Uwo8CVKx9sQ-MC34:b-896fix) ([raw file](https://github.com/gramineproject/gramine/blob/180dbc57ccabb2c18afd83925df9da45776bc8c2/libos/src/libos_async.c#L327)):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

Can we move this global assignment to the very end of the func? This will be more in line with our conventions.
</blockquote></details>

Done.
___
*[`libos/src/libos_async.c` line 356 at r2](https://reviewable.io/reviews/gramineproject/gramine/1992#-O5vrBYAEukl6XrrHKn9:-O61_l5MEKrOWP1QnMmI:b-hfpv0l) ([raw file](https://github.com/gramineproject/gramine/blob/180dbc57ccabb2c18afd83925df9da45776bc8c2/libos/src/libos_async.c#L356)):*
<details><summary><i>Previously, mkow (Michał Kowalczyk) wrote…</i></summary><blockquote>

Whatever @mwkmwkmwk decides, if choosing the first option, please add a comment explaining that this function is called when terminating the whole process so it doesn't need to clean up everything.
</blockquote></details>

removed

**Review by mkow — APPROVED** [review]
> Reviewed 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/1992)* status: all files reviewed, 4 unresolved discussions, not enough approvals from maintainers (2 more required), not enough approvals from different teams (1 more required, approved so far: ITL), "fixup! " found in commit messages' one-liners (waiting on @dimakuv)

**Review by dimakuv — APPROVED** [review]
> Reviewed 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/1992)* status: all files reviewed, all discussions resolved, not enough approvals from maintainers (1 more required), "fixup! " found in commit messages' one-liners

**Review by mkow — COMMENTED** [review]
@mwkmwkmwk: One more thing: we need your sign-off (see https://gramine.readthedocs.io/en/latest/devel/DCO/).

> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/1992#-:-O638k5VBJlUebNTrmtj:b-26uxwk)* status: all files reviewed, all discussions resolved, not enough approvals from maintainers (1 more required), "fixup! " found in commit messages' one-liners

**Review by mwkmwkmwk — COMMENTED** [review]
rebased and signed off

> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/1992#-:-O66LtXy2uEOBOoht3k9:b-5mhorr)* status: all files reviewed, all discussions resolved, not enough approvals from maintainers (1 more required)

**Review by mkow — APPROVED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/1992)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved

**Review by dimakuv — APPROVED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/1992)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved

**Pull request merged**
